### PR TITLE
Remove invalid user EIDs and UIDs from bid request

### PR DESCRIPTION
**Pull request by Pubmatic-Supriya-Patil**
This PR solves issue https://github.com/prebid/prebid-server/issues/3859

### CODE REVIEW

**Comment by Pubmatic-Supriya-Patil** [comment]
Hi @bsardo @bretg 
Please review this PR and let me know if any changes are required.

File: errortypes/code.go [hunk 1]
```diff
@@ -35,6 +35,8 @@ const (
 	InvalidBidResponseDSAWarningCode
 	SecCookieDeprecationLenWarningCode
 	SecBrowsingTopicsWarningCode
+	InvalidUserEIDsWarningCode
+	InvalidUserUIDsWarningCode
```
Comment (SyntaxNode) [inline errortypes/code.go:39]: Do these new warning codes match with PBS-Java? We try to keep in sync.

File: endpoints/openrtb2/sample-requests/invalid-whole/user-ext-eids-uids-id-empty.json [hunk 2]
```diff
@@ -1,42 +0,0 @@
-{
```
Comment (SyntaxNode) [inline endpoints/openrtb2/sample-requests/invalid-whole/user-ext-eids-uids-id-empty.json:1]: Instead of removing these tests cases, please modify them to prove the new behavior.

File: endpoints/openrtb2/auction.go [hunk 3]
```diff
@@ -1296,8 +1296,14 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
 	// Check Universal User ID
 	eids := userExt.GetEid()
 	if eids != nil {
-		eidsValue := *eids
-		for eidIndex, eid := range eidsValue {
+
```
Comment (SyntaxNode) [inline endpoints/openrtb2/auction.go:1299]: Nitpick; Please remove the leading empty line.

File: endpoints/openrtb2/auction.go [hunk 4]
```diff
@@ -1296,8 +1296,14 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
...
+		validEids, eidErrors := validateEIDs(*eids)
+
+		if len(eidErrors) > 0 {
+			errL = append(errL, eidErrors...)
+		}
```
Comment (SyntaxNode) [inline endpoints/openrtb2/auction.go:1304]: Likely fine to keep this logic here for now. We want to separate validation from fixing in the future. We can refactor this new logic along with the rest later. Thoughts @bsardo?

File: endpoints/openrtb2/auction.go [hunk 5]
```diff
@@ -1317,6 +1323,46 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
...
+	for _, eid := range eids {
+		validUIDs, uidErrors := validateUIDs(eid.UIDs)
+		errorsList = append(errorsList, uidErrors...)
+
+		if len(validUIDs) > 0 {
+			eid.UIDs = validUIDs
+			validEIDs = append(validEIDs, eid)
+		} else {
+			errorsList = append(errorsList, &errortypes.Warning{
+				Message:     fmt.Sprintf("Removed EID with empty UIDs (source: %s)", eid.Source),
```
Comment (SyntaxNode) [inline endpoints/openrtb2/auction.go:1339]: Please keep with the established error message format. Perhaps something like:

```
request.user.ext.eids[0] (source: %s) contains only empty uids and is removed from the request
```

File: endpoints/openrtb2/auction_test.go [hunk 6]
```diff
@@ -5922,3 +5922,179 @@ func sortUserData(user *openrtb2.User) {
...
+func TestValidateEIDs(t *testing.T) {
+	testCases := []struct {
+		name                  string
+		input                 []openrtb2.EID
+		expected              []openrtb2.EID
+		numErrors             int
+		expectedErrorMessages []string
+	}{
+		{
+			name: "Valid EID with non-empty UID",
```
Comment (SyntaxNode) [inline endpoints/openrtb2/auction_test.go:5934]: Go test names cannot contain spaces so they are replaced with a dash character. This makes it hard to locate the failing test based on its name. Recommend using simpler test names like:

- `one-eid-one-uid-valid`
- `one-eid-one-uid-empty`
- `many-mixed`
- `one-eid-many-uid-empty`

The parent test name is within scope so it's easy to locate in source. Please apply this guidance to other tests added in this PR.

File: endpoints/openrtb2/auction_test.go [hunk 7]
```diff
@@ -5922,3 +5922,179 @@ func sortUserData(user *openrtb2.User) {
...
+				"Removed EID with empty UIDs (source: src6)",
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			validEIDs, errorsList := validateEIDs(tc.input)
+
+			if len(validEIDs) != len(tc.expected) {
```
Comment (SyntaxNode) [inline endpoints/openrtb2/auction_test.go:5994]: Please use testify assert statements where possible. This can be written as follows which will compare the contents instead of just the array lengths, yielding more concrete tests.

```
assert.ElementsMatch(e, tc.expected, validEIDs)
```

File: endpoints/openrtb2/auction_test.go [hunk 8]
```diff
@@ -5922,3 +5922,179 @@ func sortUserData(user *openrtb2.User) {
...
+				},
+			},
+		},
+		{
+			name:              "No UIDs",
+			input:             []openrtb2.UID{},
+			expectedValidUIDs: nil,
+			expectedErrors:    nil,
+		},
+	}
```
Comment (SyntaxNode) [inline endpoints/openrtb2/auction_test.go:6089]: Please add a test case for a nil input.

**Comment by bsardo** [comment]
Hi @Pubmatic-Supriya-Patil, can you please merge with master (no rebase) and resolve conflicts? Thanks!

**Comment by SyntaxNode** [comment]
Thank you for contributing this feature Please address the comments and we will re-review.

**Comment by Pubmatic-Supriya-Patil** [comment]
> Hi @Pubmatic-Supriya-Patil, can you please merge with master (no rebase) and resolve conflicts? Thanks!

HI @bsardo I have updated branch and addressed review comments.

File: errortypes/code.go (hunk 1 above)
Comment (Pubmatic-Supriya-Patil) [reply errortypes/code.go:39]: Can you please help me to get those error codes. Do we have any document or PR to check new warning codes match with PBS-Java.

Comment (Pubmatic-Supriya-Patil) [reply errortypes/code.go:39]: As per [prebid-server-java/pull/3465](https://github.com/prebid/prebid-server-java/pull/3465), I understand that the same 999 warning code is being used when removing eids arrays. In the Prebid Server Go repository, 999 corresponds to UnknownErrorCode in the error codes reference. Are we intending to use UnknownErrorCode here as well?

In this PR, I introduced InvalidUserEIDsWarningCode (10013) and InvalidUserUIDsWarningCode (10014). If we are okay with a generic error code, we could use UnknownErrorCode (999). However, please confirm whether we want separate error codes for InvalidUserEIDs and InvalidUserUIDs, or if using 999 would suffice.

Kindly provide your input on this.

File: endpoints/openrtb2/auction.go [hunk 9]
```diff
@@ -1277,25 +1277,73 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
...
+	var validUIDs []openrtb2.UID
+	var uidErrors []error
+
+	for uidIndex, uid := range uids {
+		if uid.ID != "" {
+			validUIDs = append(validUIDs, uid)
+		} else {
+			uidErrors = append(uidErrors, &errortypes.Warning{
+				Message:     fmt.Sprintf("request.user.eids.uids[%d] contains empty ids and is removed from the request", uidIndex),
+				WarningCode: errortypes.InvalidUserUIDsWarningCode,
```
Comment (VeronikaSolovei9) [inline endpoints/openrtb2/auction.go:1339]: I used the example of user.EIDs from the issue description: 
```json
"eids": [
			{
				"source": "uidapi.com",
				"uids": [
					{"id": ""},
					{"id": "AgAACYa0cG2jcVcM42xG"}
				]
			},
			{
				"source": "liveramp.com",
				"uids": [
					{"id": "AgAACYa0cG2jcVcM42xG"}
				]
			},
			{
				"source": "yahoo.com",
				"uids": [
					{"id": ""},
					{"id": ""}
				]
			}
		]
```
As a result I received a list of errors that looked like this: 
```
request.user.eids.uids[0] contains empty ids and is removed from the request
request.user.eids.uids[0] contains empty ids and is removed from the request
request.user.eids.uids[1] contains empty ids and is removed from the request
request.user.eids[2] (source: yahoo.com) contains only empty uids and is removed from the request
```

In order to avoid confusion and duplicates may you please pass eidIndex to `validateUIDs`:
Lines 1311 and 1312:
```Golang
for eidIndex, eid := range eids {
		validUIDs, uidErrors := validateUIDs(eid.UIDs, eidIndex)
```
and then add eidIndex to error message like this: 
Line 1338:
```Golang
Message:     fmt.Sprintf("request.user.eids[%d].uids[%d] contains empty ids and is removed from the request", eidIndex, uidIndex),
```

This will give error list with no duplicate error messages: 
```
request.user.eids[0].uids[0] contains empty ids and is removed from the request
request.user.eids[2].uids[0] contains empty ids and is removed from the request
request.user.eids[2].uids[1] contains empty ids and is removed from the request
request.user.eids[2] (source: yahoo.com) contains only empty uids and is removed from the request
```

File: endpoints/openrtb2/auction.go [hunk 10]
```diff
@@ -1277,25 +1277,73 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
...
-			if uid.ID == "" {
-				return append(errL, fmt.Errorf("request.user.eids[%d].uids[%d] missing required field: \"id\"", eidIndex, uidIndex))
+			if len(eid.UIDs) == 0 {
+				return append(errL, fmt.Errorf("request.user.eids[%d].uids must contain at least one element or be undefined", eidIndex))
+			}
+
+			for uidIndex, uid := range eid.UIDs {
+				if uid.ID == "" {
+					return append(errL, fmt.Errorf("request.user.eids[%d].uids[%d] missing required field: \"id\"", eidIndex, uidIndex))
+				}
```
Comment (VeronikaSolovei9) [inline endpoints/openrtb2/auction.go:1299]: At this time we will not have empty `did.ID`, correct? there is a check for this already in line 1334.

Comment (VeronikaSolovei9) [reply endpoints/openrtb2/auction.go:1299]: Also do we need to set "clean" EIDs back to `req.User.EIDs`?

**Comment by ShriprasadM** [comment]
@Pubmatic-Supriya-Patil  : Please respond to this comments.

File: endpoints/openrtb2/auction.go (hunk 10 above)
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/auction.go:1299]: Can we remove if len(eid.UIDs) == 0 { validation also from validateUser function. This is also not possible because validateEids wil return error for emtpy uids. Please provide your suggestions @VeronikaSolovei9

**Comment by bsardo** [comment]
@Pubmatic-Supriya-Patil can you please merge with master? That should resolve the failing validate-merge check.

File: endpoints/openrtb2/auction.go (hunk 10 above)
Comment (VeronikaSolovei9) [reply endpoints/openrtb2/auction.go]: No no, I didn't ask to delete `if len(eid.UIDs) == 0 {` - it is needed here. I only asked about if we want to add valid aids to the request. And you fixed this by adding this line: `req.User.EIDs = validEids`

Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/auction.go]: Okay. Thanks. Please review PR

File: endpoints/openrtb2/auction.go (hunk 9 above)
Comment (VeronikaSolovei9) [reply endpoints/openrtb2/auction.go]: Thank you for the change!

File: endpoints/openrtb2/auction.go [hunk 11]
```diff
@@ -1277,25 +1277,69 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
...
+		for eidIndex, eid := range validEids {
+			if eid.Source == "" {
+				return append(errL, fmt.Errorf("request.user.eids[%d] missing required field: \"source\"", eidIndex))
+			}
 
-		for uidIndex, uid := range eid.UIDs {
-			if uid.ID == "" {
-				return append(errL, fmt.Errorf("request.user.eids[%d].uids[%d] missing required field: \"id\"", eidIndex, uidIndex))
+			if len(eid.UIDs) == 0 {
+				return append(errL, fmt.Errorf("request.user.eids[%d].uids must contain at least one element or be undefined", eidIndex))
```
[the remark this answers is not in the document]
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/auction.go:1293]: As per my understanding, the logic for removing invalid IDs is specific to eids. We have retained the existing logic for checking the source. However, it is unclear whether we need to check for an empty source and remove that entry while allowing other valid eids. @VeronikaSolovei9, could you please confirm this scenario and clarify the expected behavior in this case?

[the remark this answers is not in the document]
Comment (VeronikaSolovei9) [reply endpoints/openrtb2/auction.go:1293]: I will ask team about the expected behavior.

[the remark this answers is not in the document]
Comment (bretg) [reply endpoints/openrtb2/auction.go:1293]: My take is that EID.source is required. IMO, makes sense to remove empty sources.

[the remark this answers is not in the document]
Comment (bsardo) [reply endpoints/openrtb2/auction.go:1293]: The team discussed offline. We're in agreement that if `EID.source` is invalid we should simply remove the corresponding EID without failing the request. We should also add a warning message.

[the remark this answers is not in the document]
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/auction.go:1293]: Thanks for the update. I’ll incorporate the required changes.

**Review by VeronikaSolovei9 — APPROVED** [review]
Thank you for the changes, approved!

File: endpoints/openrtb2/auction.go [hunk 12]
```diff
@@ -1277,23 +1277,69 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
...
+		validEids, eidErrors := validateEIDs(req.User.EIDs)
+
+		if len(eidErrors) > 0 {
+			errL = append(errL, eidErrors...)
+		}
+		for eidIndex, eid := range validEids {
+			if len(eid.UIDs) == 0 {
+				return append(errL, fmt.Errorf("request.user.eids[%d].uids must contain at least one element or be undefined", eidIndex))
+			}
+		}
```
[the remark this answers is not in the document]
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/auction.go]: Removed

File: endpoints/openrtb2/auction_test.go [hunk 13]
```diff
@@ -5998,3 +5999,363 @@ func sortUserData(user *openrtb2.User) {
...
+				}
+
+				for i, expectedMsg := range tc.expectedErrorMessages {
+					if i >= len(errorMessages) {
+						t.Errorf("Expected error message %q but got none", expectedMsg)
+					} else if expectedMsg != errorMessages[i] {
+						t.Errorf("Expected error message %q but got %q", expectedMsg, errorMessages[i])
+					}
+				}
+			}
```
Comment (bsardo) [inline endpoints/openrtb2/auction_test.go:6090]: I think we can simplify this test quite a bit by getting rid of `numErrors` and replacing this logic as follows:
```
assert.Equal(t, len(tc.expectedErrorMessages), len(errorsList))
for _, err := range errorsList {
	assert.Contains(t, tc.expectedErrorMessages, err.Error())
}
```

File: endpoints/openrtb2/auction_test.go [hunk 14]
```diff
@@ -5998,3 +5999,363 @@ func sortUserData(user *openrtb2.User) {
...
+			expectedErrors:    nil,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			validUIDs, errorsList := validateUIDs(tc.input, 0)
+
+			assert.ElementsMatch(t, tc.expectedValidUIDs, validUIDs, "Valid UIDs mismatch")
+			assert.ElementsMatch(t, tc.expectedErrors, errorsList, "Errors mismatch")
```
Comment (bsardo) [inline endpoints/openrtb2/auction_test.go:6174]: Nitpick: you can delete the last parameter in these assert statements. `t.Run` will give the appropriate context should one of these assertions fail.

File: endpoints/openrtb2/auction.go [hunk 15]
```diff
@@ -1277,23 +1277,64 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
...
+
+func validateEIDs(eids []openrtb2.EID) ([]openrtb2.EID, []error) {
+	var errorsList []error
+	validEIDs := make([]openrtb2.EID, 0, len(eids))
+
+	for eidIndex, eid := range eids {
 		if eid.Source == "" {
-			return append(errL, fmt.Errorf("request.user.eids[%d] missing required field: \"source\"", eidIndex))
+			errorsList = append(errorsList, &errortypes.Warning{
+				Message:     fmt.Sprintf("request.user.eids[%d] missing required field: source", eidIndex),
```
Comment (bsardo) [inline endpoints/openrtb2/auction.go:1300]: Nitpick: can we change the error message a bit so that all of the messages in this function have similar wording?
`request.user.eids[%d] removed due to missing source`

File: endpoints/openrtb2/auction.go [hunk 16]
```diff
@@ -1277,23 +1277,64 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
...
+		errorsList = append(errorsList, uidErrors...)
 
-		if len(eid.UIDs) == 0 {
-			return append(errL, fmt.Errorf("request.user.eids[%d].uids must contain at least one element or be undefined", eidIndex))
+		if len(validUIDs) > 0 {
+			eid.UIDs = validUIDs
+			validEIDs = append(validEIDs, eid)
+		} else {
+			errorsList = append(errorsList, &errortypes.Warning{
+				Message:     fmt.Sprintf("request.user.eids[%d] (source: %s) contains only empty uids and is removed from the request", eidIndex, eid.Source),
```
Comment (bsardo) [inline endpoints/openrtb2/auction.go:1313]: Nitpick: can we change the error message a bit so that all of the messages in this function have similar wording and so it is a bit more concise?
`request.user.eids[%d] (%s) removed due to empty uids`

File: endpoints/openrtb2/auction.go [hunk 17]
```diff
@@ -1277,23 +1277,64 @@ func (deps *endpointDeps) validateUser(req *openrtb_ext.RequestWrapper, aliases
...
+func validateUIDs(uids []openrtb2.UID, eidIndex int) ([]openrtb2.UID, []error) {
+	var validUIDs []openrtb2.UID
+	var uidErrors []error
+
+	for uidIndex, uid := range uids {
+		if uid.ID != "" {
+			validUIDs = append(validUIDs, uid)
+		} else {
+			uidErrors = append(uidErrors, &errortypes.Warning{
+				Message:     fmt.Sprintf("request.user.eids[%d].uids[%d] contains empty ids and is removed from the request", eidIndex, uidIndex),
```
Comment (bsardo) [inline endpoints/openrtb2/auction.go:1331]: Nitpick: can we change the error message a bit so that all of the messages in this function have similar wording and so it is a bit more concise?
`request.user.eids[%d].uids[%d] removed due to empty ids`

File: endpoints/openrtb2/auction_test.go [hunk 18]
```diff
@@ -5998,3 +5999,363 @@ func sortUserData(user *openrtb2.User) {
...
+func TestValidateUser(t *testing.T) {
+
+	testCases := []struct {
+		description  string
+		req          *openrtb_ext.RequestWrapper
+		expectedErr  []error
+		expectedEids []openrtb2.EID
+	}{
+		{
+			description: "Valid user with Geo accuracy",
```
Comment (bsardo) [inline endpoints/openrtb2/auction_test.go:6187]: Nitpick: use `name` instead of `description` and replace all test case name spaces with underscores

Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/auction_test.go:6187]: Updated

File: endpoints/openrtb2/auction.go (hunk 17 above)
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/auction.go:1331]: Done

File: endpoints/openrtb2/auction.go (hunk 16 above)
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/auction.go:1313]: Done

File: endpoints/openrtb2/auction.go (hunk 15 above)
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/auction.go:1300]: Done

File: endpoints/openrtb2/auction_test.go (hunk 14 above)
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/auction_test.go:6174]: Removed last parameter

File: endpoints/openrtb2/auction_test.go (hunk 13 above)
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/auction_test.go:6090]: Done

File: endpoints/openrtb2/auction_test.go [hunk 19]
```diff
@@ -5998,3 +5999,363 @@ func sortUserData(user *openrtb2.User) {
...
 }
+func TestValidateEIDs(t *testing.T) {
+	testCases := []struct {
+		name                  string
+		input                 []openrtb2.EID
+		expected              []openrtb2.EID
+		numErrors             int
+		expectedErrorMessages []string
+	}{
+		{
```
[the remark this answers is not in the document]
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/auction_test.go:6010]: Added test cases for given scenarios

File: endpoints/openrtb2/sample-requests/invalid-whole/user-eids-uids-id-empty.json [hunk 20]
```diff
@@ -35,6 +35,6 @@
       }]
     }
   },
-  "expectedReturnCode": 400,
-  "expectedErrorMessage": "Invalid request: request.user.eids[0].uids[0] missing required field: \"id\"\n"
+  "expectedReturnCode": 200,
```
[the remark this answers is not in the document]
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/sample-requests/invalid-whole/user-eids-uids-id-empty.json:38]: Done

File: endpoints/openrtb2/sample-requests/invalid-whole/user-eids-uids-id-empty.json
[the remark this answers is not in the document]
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/sample-requests/invalid-whole/user-eids-uids-id-empty.json:1]: Done

File: endpoints/openrtb2/sample-requests/invalid-whole/user-eids-source-empty.json [hunk 21]
```diff
@@ -36,6 +36,6 @@
       }]
     }
   },
-  "expectedReturnCode": 400,
-  "expectedErrorMessage": "Invalid request: request.user.eids[0] missing required field: \"source\"\n"
+  "expectedReturnCode": 200,
+  "expectedErrorMessage": "request.user.eids[0] missing required field: source"
```
[the remark this answers is not in the document]
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/sample-requests/invalid-whole/user-eids-source-empty.json:40]: Done

**Review by VeronikaSolovei9 — APPROVED** [review]

File: endpoints/openrtb2/sample-requests/invalid-whole/user-eids-uids-id-empty.json
[the remark this answers is not in the document]
Comment (Pubmatic-Supriya-Patil) [reply endpoints/openrtb2/sample-requests/invalid-whole/user-eids-uids-id-empty.json:1]: Done

**Review by VeronikaSolovei9 — APPROVED** [review]
